### DISCUSSION TITLE
ProjectsV2 API, can't manage issue status (columns)

### DISCUSSION THREAD

**Discussion by rmusick2000** [community/community#44265] [Apps, API and Webhooks]
### Select Topic Area

Product Feedback

### Body

The API does not support creating, adding, deleting or editing a projectV2's status field (which provides columns in the board view).  A user can do this by hand, but an app can not do this programmatically.

This prevents 3rd party developers from building apps that can manage the flow of issues in a project.  The app can not create or manage a standard column format without human intervention.  Furthermore, automated testing suffers, since a testing app can not carry out basic user activity, like creating a new project column in a board view.  

Will programmatic column management be possible in 2023?

This is a blocker for many app developers.

A few relevant posts: 
https://github.com/community/community/discussions/6063
https://github.com/orgs/community/discussions/28382
 https://github.com/orgs/community/discussions/38225

**Comment by javsjmoore** [comment]
The columns are based on a field. By default, that is `Status`. You can update the status of the ProjectsV2Item with this mutation: https://docs.github.com/en/graphql/reference/mutations#updateprojectv2itemfieldvalue

**Reply by ariCETester** [reply]
Thanks for responding @javsjmoore .
I am able to change the value of the status field of a pv2 item to a pre-existing optionId (i.e. move a card between pre-existing columns).

I can not add/delete options to/from the status field (i.e. columns to/from the project) nor change the text value of an existing option.  
This means I can not create a new project and populate it with my preferred columns, nor can I edit default existing columns to change them 
into my preferred choices.  Users can of course do this by hand..

For example, I can move a card to a different column with the following, as long as the optionId (column) already exists:

mutation { 
updateProjectV2ItemFieldValue( input:{
    projectId: "PVT_kwDOA8JELs4ALvih",             
    itemId:    "PVTI_lADOA8JELs4ALvihzgFbtoA",          #..............pv2 item id, i.e. the card
    fieldId:   "PVTSSF_lADOA8JELs4ALvihzgHfSLo",     #.......status field id
    value:     {singleSelectOptionId: "f75ad846" }})           #...............optionId, i.e. the column id
    }) {clientMutationId}}

I can not, for example, modify the text for that option id (i.e. edit the name of the column).  Say I change the value input to {text: "bagels"}, then 
 - change the above mutation so itemId = status field id "PVTSSF..."            ==>  "Could not resolve to ProjectV2Item node with the global id"..
 - change the above mutation so itemId = optionID "8dc16176",                   ==>  "Could not resolve to a node with the global id"..
 - leave the project & item ids alone, but set fieldId = "optionId "8dc16176"   ==>  "Could not resolve to a node with the global id"..

As far as I can tell, neither the status field nor the optionId are projectV2 items, so this mutation can't operate on them.

Hopefully I'm missing something?

**Comment by javsjmoore** [comment]
Do these mutations match what you're looking for? 
Create field: https://docs.github.com/en/graphql/reference/mutations#createprojectv2field
Delete field: https://docs.github.com/en/graphql/reference/mutations#deleteprojectv2field
Clear field: https://docs.github.com/en/graphql/reference/mutations#deleteprojectv2field

If you have a project you've already setup, are you able to copy it with all your templated fields?
Copy ProjectV2: https://docs.github.com/en/graphql/reference/mutations#deleteprojectv2field

**Reply by ariCETester** [reply]
Thank you for the pointers!  

I can create and delete custom fields.   But when trying to operate on "Status",  I get 'Name cannot have a reserved value, Name has already been taken', or 'Only custom fields can be deleted.'

Custom fields do allow a workaround, but the workaround is sloppy for what we need to do.  Our 3rd party tool needs to fit into pre-existing repo and project environments.  Difficulties with the custom field workaround: 
 - I'm not sure if we can use the API to modify the default column field to a custom field, and even if we do, it goes against user expectations
 - We can't use the Status field, because we can't add our required additional options (columns).
 - Users can modify Status, and add/delete/edit individual values for fields.  We can't write unit tests to cover these situations.
 - Templates don't help, since we need to fit into pre-existing environments.

**Reply by chicco785** [reply]
The issue is that the mutations don't allow to update the available options to a field afaik :(

**Reply by javsjmoore** [reply]
@chicco785 You should be able to edit all the fields. Can you give an example?

**Reply by chicco785** [reply]
I am able to update all fields in an item, what I am not able to do is edit available options to a project field of type single selector. I tried using create, assuming that I could overwrite a selection field with the update version and that the API would not create new node ids for the existing option values. If I try that, it complains that the name of the field is already taken :/

```
async function createProjectField(
  projectId: string,
  name: string,
  dataType: string,
  token: string,
  singleSelectOptions?: SelectOption[],
): Promise<CreatedProjectField | undefined> {
  try {
  const octokit = new Octokit({ auth: token })
  const response = (await octokit.graphql(
    `
    mutation createProjectField($projectId: ID!, $name: String!, $dataType: ProjectV2CustomFieldType!, $singleSelectOptions: [ProjectV2SingleSelectFieldOptionInput!]) {
      createProjectV2Field(input: { projectId: $projectId, name: $name, dataType: $dataType, singleSelectOptions: $singleSelectOptions }) {
        projectV2Field {
          ... on ProjectV2Field {
            id
            name
          }
          ... on ProjectV2IterationField {
            id
            name
          }
          ... on ProjectV2SingleSelectField {
            id
            name
            options {
              id
              name
            }
          }
        }
      }
    }`,
    { projectId: projectId, name, dataType, singleSelectOptions },
  )) as { createProjectV2Field: { projectV2Field: CreatedProjectField } };

  return response.createProjectV2Field.projectV2Field;
} catch (err) {
  core.error(
    `Failed to create field configuration`
  )
}
return 
};
```

**Reply by javsjmoore** [reply]
Maybe you need to clear and then update? I will try this myself once I have some time.

https://docs.github.com/en/graphql/reference/mutations#clearprojectv2itemfieldvalue
https://docs.github.com/en/graphql/reference/mutations#updateprojectv2itemfieldvalue

**Reply by chicco785** [reply]
updating an item field is not a problem, adding a new value to available options of a project field instead it is.

surely, I could then wip a full field in a project, and re-create it, but then I will have to re-update all the items, to not only set the new option value, but also re-set the values previously set, given that the field was deleted at the project level and the new one has different node id and options node ids.

**Reply by danielmarbach** [reply]
@javsjmoore Have you had a chance to try this? I haven't been able in any way to modify the existing status field options. You can't delete the custom field and you can only clear a field that is associated with an item. So I'm kind of stuck. It seems like mentioned in the issue it is only possible to create new custom fields and use them as columns.

**Reply by javsjmoore** [reply]
Without deleting the status and adding a new one, I don't currently see a way to achieve this.

**Reply by danielmarbach** [reply]
The default status field cannot be deleted at all unfortunately

**Comment by jsiuaganatmyalii** [comment]
Hi I don't know if it some sort of related, I am using the updateIssue mutation using graphql. What I want to do is from one column inside the project is to move to the other column. Is there a way to do that via API? In UpdateIssue Mutation I can sort of change the values of the issue but the main point of moving it to the other column is I cannot figure out

**Reply by javsjmoore** [reply]
The `Column By` property in your project is based on `ProjectV2Field` values. If you go to the project's settings, you can see the `Custom fields` that define your columns.

Here's an example of a mutation that can do that:
```
        mutation changeStatus(
          $field: ID!
          $item: ID!
          $project: ID!
          $value: ProjectV2FieldValue!
        ) {
          updateProjectV2ItemFieldValue(
            input: {
              fieldId: $field
              itemId: $item
              projectId: $project
              value: $value
            }
          ) {
            clientMutationId
            projectV2Item {
              id
            }
          }
        }
```

The `$value` will be some sort of [ProjectV2FieldValue](https://docs.github.com/en/graphql/reference/input-objects#projectv2fieldvalue).

Here's a link to the mutation that GraphQL is calling: https://docs.github.com/en/graphql/reference/input-objects#updateprojectv2itemfieldvalueinput.

In V1 Projects, you had columns that contained cards. Now, you can think of it as every card having fields and the views group the columns based on fields.

**Comment by LipJ01** [comment]
I am struggling to figure out what the fieldId is. Any suggestions on how to obtain it using the graphQl API?

**Reply by javsjmoore** [reply]
Something like this might work. Variables at the bottom. Change `MyOrg` and `MyProject` to whatever yours are.
```
query getProjectStatuses($org: String!, $project: String!) {
  organization(login: $org) {
    projectsV2(query: $project, first: 1) {
      edges {
        node {
          fields(first: 100) {
            edges {
              node {
                ...on ProjectV2FieldCommon {
                  name,	
                  id
                }
              }
            }
          }
        }
      }
    }
  }
}
```

```
{
  "org": "MyOrg",
  "project": "MyProject"
}
```

**Comment by NerdGGuy** [comment]
Some kind of "updateProjectV2Field" where you can specify the new field names `updateProjectV2Field(input: {..., singleSelectOptions: $singleSelectOptions})` is what I'm after 🤞.
